Show breakdown of total issues in panelFooter #1377

yl-coder opened the pull request:
Ready for review
fixes #1377 

![image](https://cloud.githubusercontent.com/assets/9070608/13899064/e4505180-edba-11e5-87f6-7d073b86b23c.png)

damithc commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -144,6 +166,9 @@ public final void refreshItems() {
         issueCount = getElementsList().size();
+        closedIssueCount = getClosedIssueSize();
+        openIssueCount = issueCount - closedIssueCount;
+        System.out.println("closed = " + closedIssueCount + " open = " + openIssueCount);
```
May be we use this format instead `T (O + C)`. If Total is 0, we can omit the part inside brackets.

yl-coder replied:
Yes agree, implemented it. it looks better now.

yl-coder commented:
Ready for review

ndt93 commented:
Can you rebase the current master into this? There are a recent changes that make the tests more stable.

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -35,19 +39,29 @@
-    private int issueCount;
+    private int issueCount = 0;
+    private int closedIssueCount = 0;
+    private int openIssueCount = 0;
```
These `*counts` variables should have plural form? e.g. `issuesCount`

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -12,13 +13,16 @@
+import javafx.scene.paint.Color;
+
 
+import javafx.scene.text.TextFlow;
```
Not sure if your IDE settings create this but 1 blank line should be enough. Can also your IDE's automatic imports organization for these stuffs.

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -35,19 +39,29 @@
-    private Label totalLabel;
+
+
+    Text openIssueText;
```
Same thing there. I think 1 blank line should be readable enough to separate groups of variables.

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +181,39 @@ public final void refreshItems() {
     }
 
+    private int getClosedIssueSize() {
+        int count = 0;
```
Should be `getClosedIssuesCount`? To be consistent of the rest of variable names.

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +181,39 @@ public final void refreshItems() {
+            }
+        }
+        return count;
+    }
```
If you are familiar with Java 8's functional stuffs, you can do this in a more functional way like this:
```java
return getElementsList().stream().filter((element) -> !element.getIssue().isOpen()).count();
```

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +181,39 @@ public final void refreshItems() {
+            openIssueText.setText(String.valueOf(openIssueCount));
+            closedIssueText.setText(String.valueOf(closedIssueCount));
+            totalIssueText.setText(String.valueOf(issueCount));
+            hideFooterPanelStatsDetails(false);
```
Should be `DRYer` to extract a method to update the texts of all 3 stats text nodes, as they should be updated together (either all or none). So I think it's not a good idea to just update the `totalIssueText` like in the `if` branch.

ndt93 commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +181,39 @@ public final void refreshItems() {
+    }
+
+    private void hideFooterPanelStatsDetails(boolean hide) {
+        openIssueText.setVisible(!hide);
```
Not sure if it's a good idea to have a `hideFooter...` method and pass in a boolean to either `hide` or `show`. @damithc

ndt93 commented:
@damithc No more comments from me.

damithc commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +179,36 @@ public final void refreshItems() {
+        totalIssueText.setText(String.valueOf(issuesCount));
+    }
+
+    private void hideFooterPanelStatsDetails(boolean hide) {
```
boolean variables should be named as such. e.g. `isHidden`

damithc commented on src/main/java/ui/listpanel/ListPanel.java:
```diff
@@ -152,11 +179,36 @@ public final void refreshItems() {
 
+    private int getClosedIssuesCount() {
+        return (int) getElementsList().stream().filter((element) -> !element.getIssue().isOpen()).count();
+    }
```
@ndt93 is there a better place for this function? Generating the stats should be done closer to the data rather than the UI?

ndt93 replied:
There are 2 levels of data.

The lower level data would be in `MultiModel` i.e. `Logic#models`, suitable for stats across repositories or in a `Model` in `models` if the stats concern only 1 repository. This might not be a suitable place for the method as a panel displays mixed subsets of issues across multiple repositories depending on that panel's filter.

The higher level data is a `List` of `GuiElements` that represents the data-state of a panel and is stored within the `ListPanel` object itself. Ideally, we can have a class that encapsulate this, which would become a good place to put this method.

damithc replied:
Yes, that kind of UI-data separation can be useful. We should do that at  some point.

yl-coder replied:
I was thinking whether we should move it to FilterPanel.java now, as it is where the List of GuiElements are now.

damithc commented:
@yl-coder one minor change requested.

dariusf commented:
Looks good. @yl-coder could you squash the noisy commits? Things like fixing violations, things changed in review, etc.

yl-coder commented:
@dariusf How do i squash the commits when the commits have already been pushed to the remote? After rebase i do a force push? The previous time i did it, after rebase It says there's new commits to be pull, and after pulling everything screw up :p

dariusf commented:
Yes, because rebasing creates new commits, you will get something like "x ahead, x behind". You can force push after you check that everything went well.

(Just in case, you're at 7cc8721 now.)

yl-coder commented:
@dariusf Sorry did i do it correctly?

dariusf commented:
Doesn't look right, since there are others' commits in here now. Have you updated your fork to the latest master? Anyway, you can `git reset --hard 7cc8721` to get back to where you were. No need to make a new PR.

yl-coder commented:
I updated my fork to the latest master. It doesn't work also. It still show up others commit when i push my squash commits. i have reverted back. Is my first time squashing commits :(

damithc commented:
@yl-coder you may need to sync your branch with the latest master.

ndt93 commented:
@yl-coder You can try `rebase` master into your branch first next time instead of `merge` since `merge` will interleave other commits and yours and make squashing or cleaning up commits difficult. And we have to `git push --force` after rebase, squash, amend etc. commits which had been pushed to remote since the commits history is rewritten. Ideally that shouldn't be done but since our work flow has only 1 person on a branch so it's still generally acceptable.

@dariusf So can we merge the commits here as it is?

ndt93 commented:
Merged in ed79899
